### PR TITLE
FIR: Make FirAnnotationArgumentChecker a FirAnnotationCallChecker

### PR DESCRIPTION
**Pull request by punzki**
...to run on all annotation calls, including those inside type annotations.

### CODE REVIEW

**Review by tgeng — APPROVED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt [hunk 1]
```diff
@@ -14,19 +13,17 @@ import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
...
+object FirAnnotationArgumentChecker : FirAnnotationCallChecker() {
+    override fun check(expression: FirAnnotationCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        expression.argumentMapping?.let {
+            for ((arg, _) in it) {
+                val argExpression = (arg as? FirNamedArgumentExpression)?.expression ?: arg
 
-                checkAnnotationArgumentWithSubElements(expression, context.session, reporter, context)
-                    ?.let { reporter.reportOn(expression.source, it, context) }
+                checkAnnotationArgumentWithSubElements(argExpression, context.session, reporter, context)
+                    ?.let { reporter.reportOn(argExpression.source, it, context) }
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt:26]: It is out of scope of this PR, and not sure you will keep working on this diagnostic, but according to [this](https://punzki.github.io/Kotlin-DiagnosticTestsDiff/2021-05-05/diffs/tests/varargs/assignNonConstSingleArrayElementAsVarargInAnnotationError.kt.html), I guess this source needs some fixes.

**Review by jsjeon — APPROVED** [review]

**Review by mglukhikh — APPROVED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt [hunk 2]
```diff
@@ -14,19 +13,17 @@ import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
...
-object FirAnnotationArgumentChecker : FirBasicDeclarationChecker() {
-    override fun check(declaration: FirDeclaration, context: CheckerContext, reporter: DiagnosticReporter) {
-        if (declaration !is FirAnnotationContainer) return
-        for (declarationOfAnnotation in declaration.annotations) {
-            for ((arg, _) in declarationOfAnnotation.argumentMapping ?: continue) {
-                val expression = (arg as? FirNamedArgumentExpression)?.expression ?: arg
+object FirAnnotationArgumentChecker : FirAnnotationCallChecker() {
+    override fun check(expression: FirAnnotationCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        expression.argumentMapping?.let {
+            for ((arg, _) in it) {
```
[the remark this answers is not in the document]
Comment (mglukhikh) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt:22]: I'll cleanup this place via `val argumentMapping = expression.argumentMapping ?: return` at first line

**Comment by mglukhikh** [comment]
Merged manually 👍

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (punzki) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt:22]: Thanks @mglukhikh, that works.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt (hunk 1 above)
Comment (punzki) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt:26]: Thanks for pointing that out (I should use my own data :slightly_smiling_face:). I think those are fixes on source positioning side.

**Pull request closed without merging**